sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,333 @@
+    if (endpoint.isTimer()) {
+      return registerOutputTimersLocation(
+          endpoint.getTransformId(), endpoint.getTimerFamilyId(), coder);
+    } else {
+      return registerOutputDataLocation(endpoint.getTransformId(), coder);
+    }
+  }
+
+  public void flush() throws IOException {
+    Elements.Builder elements;
+    elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    if (totalByteCounter == 0 && outputDataCoders.isEmpty() && outputTimersCoders.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    LOG.debug(
+        "Closing streams for outbound endpoints {} and {}.",
+        outputDataCoders.keySet(),
+        outputTimersCoders.keySet());
+    LOG.debug(
+        "Sent outbound data size : {}, outbound timers size : {}.",
+        perEndpointDataByteCount,
+        perEndpointTimersByteCount);
+    LOG.debug("Sent outbound element count : {}.", perEndpointElementCount);
+    for (String pTransformID : outputDataCoders.keySet()) {
+      bufferedElements
+          .addDataBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(pTransformID)
+          .setIsLast(true);
+    }
+    for (KV<String, String> pTransformAndTimerId : outputTimersCoders.keySet()) {
+      bufferedElements
+          .addTimersBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(pTransformAndTimerId.getKey())
+          .setTimerFamilyId(pTransformAndTimerId.getValue())
+          .setIsLast(true);
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    perEndpointDataByteCount.values().forEach(LongAdder::reset);
+    perEndpointTimersByteCount.values().forEach(LongAdder::reset);
+    perEndpointElementCount.clear();
+  }
+
+  public <T> void accept(
+      ByteString.Output output, Coder<T> coder, LongAdder perEndpointByteCount, T data)
+      throws Exception {
+    if (timeLimit > 0) {
+      checkFlushThreadException();
+    }
+    int size = output.size();
+    coder.encode(data, output);
+    if (output.size() - size == 0) {
+      output.write(0);
+    }
+    final long delta = (long) output.size() - size;
+    totalByteCounter += delta;
+    perEndpointByteCount.add(delta);
+
+    if (totalByteCounter >= sizeLimit) {
+      flush();
+    }
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (Map.Entry<String, ByteString.Output> entry : dataBuffer.entrySet()) {
+      if (entry.getValue() == null || entry.getValue().size() == 0) {
```

Comment:
this should never be null
also prefer isEmpty over size == 0

Reply:
Done.

Reply:
Done.